### PR TITLE
[fully_async] fix: preserve per-iteration routed_experts on partial rollout resume

### PR DESCRIPTION
**Pull request by NoonePauseferg**
### What does this PR do?

Fixes `routed_experts` merging during partial rollout resume in fully async training.

sglang returns `routed_experts` for the full sequence (prompt + all tokens). The previous fix (#6030) replaced `routed_experts` entirely on each resume iteration. However, in fully async training the model weights can update between partial rollout iterations, so the latest call's routing reflects the new model — not the one that actually generated earlier tokens.

This PR instead keeps each iteration's routing for the tokens it actually generated by slicing `output.routed_experts[-len(output.token_ids):]` and concatenating.

Follows up on #6030.

### Checklist Before Starting

- [x] Search for similar PRs: [routed_experts partial rollout](https://github.com/verl-project/verl/pulls?q=is%3Apr+routed_experts+partial+rollout) — #6030 was merged with simple replace; this PR improves on it.
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

Tested on 64-expert 10B MoE model (sglang rollout + megatron actor, R3 mode, async with partial rollout):

- **Simple replace** (red): `actor/ppo_kl` spikes up to 2.8e-4 in early training, volatile
- **Per-iteration routing** (blue): smoother `ppo_kl`, stays below 2.0e-4

recap_vs_simple

- red - before fix 
- blue - after fix

### API and Usage Example

No API changes. Existing partial rollout + R3 router replay config works as before:

```bash
async_training.partial_rollout=True
actor_rollout_ref.actor.megatron.router_replay.mode=R3
actor_rollout_ref.rollout.enable_rollout_routing_replay=True
```

### Design & Code Changes

Single file change in `verl/experimental/fully_async_policy/agent_loop/agent_loop.py`:

1. On first partial rollout iteration (`final_output.routed_experts is None`): take full `output.routed_experts` (covers prompt + first batch of tokens)
2. On subsequent iterations: slice only newly generated tokens' routing (`output.routed_experts[-len(output.token_ids):]`) and concatenate with existing
3. Guard with `len(output.token_ids) > 0` to avoid `[-0:]` slicing bug (would select entire tensor)
4. Re-add `import torch` for `torch.cat`

**Why not simple replace?** The actor update runs with the latest weights (v2). Theoretically v2 routing + v2 weights gives a cleaner π_v2. But empirically, preserving per-iteration routing gives smoother KL — the routing signal from the generating model matters more than weight-routing consistency.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always` — all passed.
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs). — N/A, no doc changes needed.
- [ ] Add unit or end-to-end test(s) — not feasible: requires MoE model + async partial rollout + multi-GPU. Validated via experiment on 64-expert 10B model.
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1).

### CODE REVIEW

File: verl/experimental/fully_async_policy/agent_loop/agent_loop.py [hunk 1]
```diff
@@ -95,10 +96,16 @@ async def generate(
...
+            # On partial rollout resume the model version may differ, so keep
+            # existing routing and only append routing for newly generated tokens.
+            if output.routed_experts is not None and len(output.token_ids) > 0:
+                if final_output.routed_experts is None:
+                    final_output.routed_experts = output.routed_experts
+                else:
+                    final_output.routed_experts = torch.cat(
+                        [final_output.routed_experts, output.routed_experts[-len(output.token_ids):]],
+                        dim=0,
+                    )
```
[the remark this answers is not in the document]
Comment (NoonePauseferg) [reply verl/experimental/fully_async_policy/agent_loop/agent_loop.py:108]: routed_experts is always a torch tensor here — both sglang (extract_routed_experts_from_meta_info().reshape()) and vllm (final_res.outputs[0].routed_experts) return tensors

**Comment by wuxibin89** [comment]
@NoonePauseferg Please format code.

**Review by wuxibin89 — APPROVED** [review]

**Comment by wuxibin89** [comment]
@NoonePauseferg Besides actor/ppo_kl, what about other metrics? critic/reward, val/reward, etc.

**Comment by wuxibin89** [comment]
@NoonePauseferg There's still problem for multi-turn rollout, since in each turn the routed_experts is overrided by the newest version, while previous turn maybe generated by older version.
https://github.com/verl-project/verl/blob/main/verl/experimental/agent_loop/tool_agent_loop.py#L265
